### PR TITLE
support access processed components in cue template context

### PR DESCRIPTION
**Pull request by captainroy-hy**
Signed-off-by: roywang <royyuewang@tencent.com>

**What this PR does / why we need it**:

Support to access components already processed before in the latter ones' cue template.
Since Policy definition is processed after components, so it also supports to access all components in Policy definition template.

Here's an PolicyDefinition that can render a HealthScope which contains workload references.
This PR helps to get workload references and add it to HealthScope `.spec.appReferences.[appName].compReferences[compName].workload`.

```cue
"scope-binding": {
	annotations: {}
	attributes: {}
	description: "Bind components to scopes"
	labels: {}
	type: "policy"
}

template: {

	output: {
		apiVersion: "core.oam.dev/v1alpha2"
		kind:       "HealthScope"
		spec: {
...
			appReferences: [{
				appName: context.appName
				compReferences: [
					for k, v in parameter.boundComponents
					if context.artifacts[v].ready {
						compName: v
						workload: {
							apiVersion: context.artifacts[v].workload.apiVersion
							kind:       context.artifacts[v].workload.kind
							name:       v
						}
					},
				]
			}]
		}
	}
	parameter: {
		boundComponents: [...string]
		probeTimeout:  *10 | int
		probeInterval: *30 | int
	}
}
```
```yaml
apiVersion: core.oam.dev/v1beta1
kind: Application
metadata:
  name: policy-test
  namespace: default
spec:
  components:
  - name: express-server
    type: webservice
    properties:
      image: crccheck/hello-world
      port: 8000
    traits:
    - type: ingress
      properties:
        domain: testsvc.example.com
        http:
          /: 8000
  - name: nginx-server
    type: webservice
    properties:
      image: nginx:1.21
      port: 80
  policies:
    - name: bind-health-scope
      type: scope-binding
      properties:
        boundComponents:
          - express-server
          - nginx-server
  workflow:
    steps:
      - name: apply-express-server
        type: apply-component
        properties:
          component: express-server
      - name: apply-nginx-server
        type: apply-component
        properties:
          component: nginx-server
```

### CODE REVIEW

**Comment by captainroy-hy** [comment]
@leejanee Pls help to review. And I will add tests if you're okay with this implementation.

**Comment by leejanee** [comment]
Improve test coverage

File: pkg/cue/process/handle.go [hunk 1]
```diff
@@ -240,9 +247,37 @@ func (ctx *templateContext) BaseContextFile() string {
 	if ctx.outputSecretName != "" {
 		buff += fmt.Sprintf("%s:\"%s\"", OutputSecretName, ctx.outputSecretName)
 	}
+
+	if len(ctx.processedComponents) > 0 {
+		processedComp := ctx.ProcessedContextFile()
+		buff += "\n" + ContextProcessedComponents + ": {" + processedComp + "\n} \n"
```
Comment (leejanee) [inline pkg/cue/process/handle.go:253]: Should be covered by tests

File: pkg/cue/process/handle.go [hunk 2]
```diff
@@ -240,9 +247,37 @@ func (ctx *templateContext) BaseContextFile() string {
...
+// ProcessedContextFile adopts processed components into current template context
+func (ctx *templateContext) ProcessedContextFile() string {
+	buff := ""
+	for _, pComp := range ctx.processedComponents {
+		if pComp.InsertConfigNotReady {
+			buff += fmt.Sprintf("\n \"%s\": ready: false", pComp.Name) + "\n"
+			continue
+		}
+		workloadData, _ := pComp.StandardWorkload.MarshalJSON()
+		traitsDataBuff := ""
```
Comment (leejanee) [inline pkg/cue/process/handle.go:267]: test case

File: pkg/cue/process/handle.go [hunk 3]
```diff
@@ -240,9 +247,37 @@ func (ctx *templateContext) BaseContextFile() string {
...
+		for _, trait := range pComp.Traits {
+			traitData, _ := trait.MarshalJSON()
+			traitsDataBuff += fmt.Sprintf("\"%s\": \"%s\": %s\n",
+				trait.GetLabels()[oam.TraitTypeLabel],
+				trait.GetLabels()[oam.TraitResource],
+				string(traitData))
+		}
+		buff += fmt.Sprintf("\n \"%s\": {\nready: true\n workload: %s\n traits: {\n%s}\n}",
+			pComp.Name, string(workloadData), traitsDataBuff)
```
Comment (leejanee) [inline pkg/cue/process/handle.go:276]: test case

**Review by leejanee — APPROVED** [review]
lgtm

**Review by wonderflow — APPROVED** [review]

**Pull request merged**